colossalai/zero/gemini/gemini_optimizer.py

```diff
@@ -301,25 +308,365 @@ def get_range_pair(local_chunk: Chunk, local_param: Parameter):
                 fake_param = torch.nn.Parameter(torch.empty([0], device=grad_device))
                 self.param_to_chunk32[fake_param] = chunk16.paired_chunk
                 self.param_to_range[fake_param] = range_pair
-
+                self.id_to_fake_params[param_id] = fake_param
                 fake_params_list.append(fake_param)
 
+            # Update self.optim.param_groups as well as backup group.
             group['params'] = fake_params_list
+            group_backup['params'] = group_ids
+            self.param_groups_backup.append(group_backup)
+
+    def get_offsets(self, param_id: int) -> tuple:
+        # Args: param_id(int): The id of parameter.
+        #
+        # Returns: chunk_offset(int):Offset of parameter inside the chunk.
+        #          shard_offset(int): Offset of its optimizer state shard
+        #                             relative to the whole optimizer state.
+        #          param_size(int): Length of parameter shard owned by current process.
+
+        assert param_id in self.id_to_fake_params
+        fake_param = self.id_to_fake_params[param_id]
+        chunk = self.param_to_chunk32[fake_param].paired_chunk
+        param = self.id_to_real_params[param_id]
+        param_info = chunk.tensors_info[param]
+
+        param_range = self.param_to_range[fake_param]
+        begin_in_chunk, end_in_chunk = param_range
+        chunk_offset = begin_in_chunk
+        shard_offset = begin_in_chunk + chunk.shard_begin - param_info.offset
+        param_size = end_in_chunk - begin_in_chunk
+        assert chunk_offset >= 0 and shard_offset >= 0
+
+        return chunk_offset, shard_offset, param_size
+
+    def collect_states(self, param_id: int) -> dict:
+        """
+        Args:
+            param_id (int): id of the parameter whose state is to be gathered at master rank.
+
+        Returns:
+            collected_states(dict): the gathered optimzier state of parameter with given id
+                                    if this method is called by master rank, otherwise an empty dict.
+
+        This method can work only when called by all processes simultaneously.
+        """
+
+        # Get param & chunk & process group.
+        param = self.id_to_real_params[param_id]
+        chunk = self.chunk_manager.get_chunk(param)
+        process_group = chunk.torch_pg
+        rank = dist.get_rank(process_group)
+        master_rank = 0
+        state_names = self.optim.get_state_names()
+
+        collected_states = {}
+
+        # If the chunk is kept gathered,
+        # the parameteres are treated the same as that of those in strict DDP during training.
+        # So states can be directly fetched from current device.
+        if chunk.keep_gathered:
+            assert param_id in self.id_to_fake_params
+            fake_param = self.id_to_fake_params[param_id]
+            if rank == master_rank:
+                states = self.optim.state[fake_param]
```

Comment:
This `states` is a dict whose key is `state_name`. Can we just get state name through it directly?

Reply:
The master rank might not own the states of a given parameter, in which case states will be empty. So I added get_state_names() method in optimizer classes to fetch names of states.

Reply:
But the state is homogeneous for all parameters. You can gather the state names before gathering states.

Reply:
Hard coded state names is hard to extend

Reply:
OK, I will try to all_gather state names

Reply:
Resolved